### PR TITLE
🏛 chore: upgrade jest, fix config

### PR DESCRIPTION
**Pull request by estrattonbailey**
Fixes RNBW-4388

## What changed (plus any additional context for devs)
Upgrades to Jest 28. The latest is 29, but it was released a week ago and `ts-jest` doesn't yet support it. Detox may be similar. `ts-jest` gives us type checking for our `.ts` files.

I moved the Detox and Jest configs into their own JS files so that they're easier to work with (can add comments, import type definitions and other utils).

Based on my testing, the `transform` option needs to vary slightly between the two Jest configs.

## What to test
- run e2e, if it passes we should be good
- other Jest tests in the codebase haven't been run in some time, so they don't matter right now — I'll fix this next

## Final checklist

- [x] Assigned individual reviewers?
- [x] Added labels? (`team1/team2`, `critical path`, `release`, `dev QA`)
- [x] Did you test both iOS and Android?
- [x] If your changes are visual, did you check both the light and dark themes?
- [x] Added e2e tests? If not, please specify why
- [x] If you added new critical path files, did you update the CODEOWNERS file?
- [x] If no `dev QA` label, did you add the PR to the QA Queue?

### CODE REVIEW

File: e2e/jest.e2e.config.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
```
Comment (estrattonbailey) [inline e2e/jest.e2e.config.js:1]: Adds type checking to this file! Very nice. This is how `ts-jest` sets it up by default.

File: e2e/jest.e2e.config.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+const { pathsToModuleNameMapper } = require('ts-jest');
```
Comment (estrattonbailey) [inline e2e/jest.e2e.config.js:2]: This is handy, means we don't have to manually keep these in sync.

File: jest.config.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
+const { pathsToModuleNameMapper } = require('ts-jest');
+const { compilerOptions } = require('./tsconfig');
+
+module.exports = {
+  preset: 'react-native',
+  setupFiles: ['./config/test/jest-setup.js'],
+  transform: {
+    '\\.tsx?$': 'ts-jest',
```
Comment (estrattonbailey) [inline jest.config.js:9]: Slight variation from the e2e config to allow other files to fall back to default Jest handling that's configured in the `react-native` preset. In Detox we can't set this preset (doesn't work), so we defer everything to `ts-jest`. Appears to work just fine.

**Review by skylarbarrera — APPROVED** [review]
🏛️

**Review by estebanmino — COMMENTED** [review]
i'm running e2e locally and is running fine but i don't see any indication of which test is running


this is very useful while trying to fix a failing test, now we won't know in which test step the e2e

**Comment by estrattonbailey** [comment]
Good call! I bet there's some config I missed. Will fix this 👍

**Review by estebanmino — APPROVED** [review]
ok if a test fails the step will be shown to devs anyways, so 🤝

**Pull request merged**